RecoLocalTracker/SiPixelRecHits/src/SiPixelTemplate2D.cc

```diff
@@ -1,801 +1,1498 @@
+   
+   // Local variables
+   int i, j, k, l, iy, jx;
+   const char *tempfile;
+   //	char title[80]; remove this
+   char c;
+   const int code_version={21};
+   
+   
+   
+   //  Create a filename for this run
+   
+   std::ostringstream tout;
+   
+   //  Create different path in CMSSW than standalone
+   
+#ifndef SI_PIXEL_TEMPLATE_STANDALONE
+   tout << dir << "template_summary2D_zp"
+   << std::setw(4) << std::setfill('0') << std::right << filenum << ".out" << std::ends;
+   std::string tempf = tout.str();
+   edm::FileInPath file( tempf.c_str() );
+   tempfile = (file.fullPath()).c_str();
+#else
+   tout << "template_summary2D_zp" << std::setw(4) << std::setfill('0') << std::right << filenum << ".out" << std::ends;
+   std::string tempf = tout.str();
+   tempfile = tempf.c_str();
+#endif
+   
+   //  open the template file
+   
+   std::ifstream in_file(tempfile, std::ios::in);
+   
+   if(in_file.is_open()) {
+      
+      // Create a local template storage entry
+      
+      SiPixelTemplateStore2D theCurrentTemp;
+      
+      // Read-in a header string first and print it
+      
+      for (i=0; (c=in_file.get()) != '\n'; ++i) {
+         if(i < 79) {theCurrentTemp.head.title[i] = c;}
+      }
+      if(i > 78) {i=78;}
+      theCurrentTemp.head.title[i+1] ='\0';
+      LOGINFO("SiPixelTemplate2D") << "Loading Pixel Template File - " << theCurrentTemp.head.title << ENDL;
+      
+      // next, the header information
+      in_file >> theCurrentTemp.head.ID  >> theCurrentTemp.head.templ_version >> theCurrentTemp.head.Bfield >> theCurrentTemp.head.NTy >> theCurrentTemp.head.NTyx >> theCurrentTemp.head.NTxx
+      >> theCurrentTemp.head.Dtype >> theCurrentTemp.head.Vbias >> theCurrentTemp.head.temperature >> theCurrentTemp.head.fluence >> theCurrentTemp.head.qscale
+      >> theCurrentTemp.head.s50 >> theCurrentTemp.head.lorywidth >> theCurrentTemp.head.lorxwidth >> theCurrentTemp.head.ysize >> theCurrentTemp.head.xsize >> theCurrentTemp.head.zsize;
+      
+      if(in_file.fail()) {LOGERROR("SiPixelTemplate2D") << "Error reading file 0A, no template load" << ENDL; return false;}
+      
+      if(theCurrentTemp.head.templ_version > 17) {
+         in_file >> theCurrentTemp.head.ss50 >> theCurrentTemp.head.lorybias
+         >> theCurrentTemp.head.lorxbias >> theCurrentTemp.head.fbin[0]
+         >> theCurrentTemp.head.fbin[1] >> theCurrentTemp.head.fbin[2];
+         
+         if(in_file.fail()) {LOGERROR("SiPixelTemplate2D") << "Error reading file 0B, no template load"
+            << ENDL; return false;}
+      } else {
+         theCurrentTemp.head.ss50 = theCurrentTemp.head.s50;
+         theCurrentTemp.head.lorybias = theCurrentTemp.head.lorywidth/2.f;
+         theCurrentTemp.head.lorxbias = theCurrentTemp.head.lorxwidth/2.f;
+         theCurrentTemp.head.fbin[0] = 1.5f;
+         theCurrentTemp.head.fbin[1] = 1.00f;
+         theCurrentTemp.head.fbin[2] = 0.85f;

```

Comment:
these magic numbers show up twice (with slightly different spelling). It looks like making a helper function would be useful.
Also, are the values well documented?

Reply:
We discussed it already and I promised to discuss this with Morris so that we can fix them eventually.  (These are the bin boundaries for Qbin distribution.)